arcgis-ios-sdk-samples/Maps/Browse building floors/BrowseBuildingFloorsViewController.swift

```diff
@@ -0,0 +1,115 @@
+// Copyright 2021 Esri
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import UIKit
+import ArcGIS
+
+class BrowseBuildingFloorsViewController: UIViewController {
+    // MARK: Storyboard views
+    
+    /// The map view managed by the view controller.
+    @IBOutlet var mapView: AGSMapView! {
+        didSet {
+            mapView.map = makeMap()
+        }
+    }
+    /// The floor level picker view.
+    @IBOutlet var floorLevelPickerView: UIPickerView!
+    
+    // MARK: Properties
+    
+    /// The levels of the floor-aware web map, with an addition of `nil`.
+    var levels: [AGSFloorLevel?] = [nil]
+    
+    // MARK: Methods
+    
+    /// Set the selected level to visible and all other levels to invisible.
+    /// - Parameter selectedLevel: The `AGSFloorLevel` selected.
+    func setVisibleFloorLevel(_ selectedLevel: AGSFloorLevel?) {
+        levels.forEach { level in
+            level?.isVisible = level == selectedLevel
+        }
+    }
+    
+    func setPickerViewLayout() {
+        floorLevelPickerView.translatesAutoresizingMaskIntoConstraints = false
+        let margin: CGFloat = 10.0
+        floorLevelPickerView.layer.cornerRadius = margin
+        NSLayoutConstraint.activate([
+            floorLevelPickerView.widthAnchor.constraint(equalToConstant: 120.0),
+            floorLevelPickerView.bottomAnchor.constraint(equalTo: mapView.attributionTopAnchor, constant: -margin),
+            floorLevelPickerView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -margin)
+        ])
+    }
+    
+    /// Create a map.
+    func makeMap() -> AGSMap {
+        // A floor-aware web map for floors of Esri Building L in Redlands.
+        let map = AGSMap(item: AGSPortalItem(
+            portal: .arcGISOnline(withLoginRequired: false),
+            itemID: "f133a698536f44c8884ad81f80b6cfc7"
+        ))
+        
+        map.load { [weak self] error in
+            guard error == nil, let self = self,
+                  // The floor manager of the web map, which exposes the sites,
+                  // facilities, and levels of the floor-aware data model.
+                  let floorManager = map.floorManager else { return }
+            
+            floorManager.load { error in
```

Comment:
This closure is strongly capturing `self`, which it shouldn't be. Since it is often hard to track that sort of thing through nested closures, what I have done and what I would recommend doing is delegate each completion handler to a method. That way, `self` only needs to be unwrapped through optional chaining. It might look something like this:

```swift
func makeMap() -> AGSMap {
    // A floor-aware web map for floors of Esri Building L in Redlands.
    let map = AGSMap(item: AGSPortalItem(
        portal: .arcGISOnline(withLoginRequired: false),
        itemID: "f133a698536f44c8884ad81f80b6cfc7"
    ))
    
    map.load { [weak self] error in
        self?.map(map, didLoadWith: error)
    }
    
    return map
}

func map(_ map: AGSMap, didLoadWith loadError: Error?) {
    guard loadError == nil else {
        return
    }
    
    // The floor manager of the web map, which exposes the sites,
    // facilities, and levels of the floor-aware data model.
    guard let floorManager = map.floorManager else {
        return
    }
    
    floorManager.load { [weak self] error in
        self?.floodManager(floorManager, didLoadWith: error)
    }
}

func floodManager(_ floorManager: AGSFloorManager, didLoadWith loadError: Error?) {
    guard loadError == nil else {
        return
    }
    
    guard let geometry = floorManager.sites.first?.geometry else {
        return
    }
    
    mapView.setViewpointGeometry(geometry)
    
    // Update floor picker and select the first floor.
    floorLevels = floorManager.levels
    let firstFloorIndex = floorLevels.firstIndex { $0.longName == "Level 1" }!
    selectedFloorLevel = floorLevels[firstFloorIndex]
}
```

You also might consider at least printing the load errors.

Reply:
> strongly capturing self, which it shouldn't be.

My bad. I mis-thought the lifespan of map view should be the same as `floorManager`, so strongly captured.